lib/client.js

```diff
@@ -2060,6 +2066,145 @@ MatrixClient.prototype.paginateEventContext = function(eventContext, opts) {
+        // have known state.
+        var events = res.events_before
+            .concat([res.event])
+            .concat(res.events_after);
+        events.reverse();
+        var matrixEvents = utils.map(events, self.getEventMapper());
+
+        var timeline = room.getTimelineForEvent(matrixEvents[0].getId());
+        if (!timeline) {
+            timeline = room.createTimeline();
+            timeline.initialiseState(utils.map(res.state,
+                                               self.getEventMapper()));
+            timeline.getState(false).paginationToken = res.end;
+        }
+        room.addEventsToTimeline(matrixEvents, true, timeline, res.start);
+
+        // there is no guarantee that the event ended up in "timeline" (we
+        // might have switched to a neighbouring timeline) - so check the
+        // room's index again. On the other hand, there's no guarantee the
+        // event ended up anywhere, if it was later redacted, so we just
+        // return the timeline we first thought of.
+        return room.getTimelineForEvent(eventId) || timeline;
+    });
+    return promise;
+};
+
+
+/**
+ * Take an EventTimeline, and back/forward-fill results.
+ *
+ * @param {module:models/event-timeline~EventTimeline} eventTimeline timeline
+ *    object to be updated
+ * @param {Object =} opts
+ * @param {boolean}  [opts.backwards = false]  true to fill backwards,
+ *    false to go forwards
+ * @param {number}   [opts.limit = 30]         number of events to request
+ *
+ * @return {module:client.Promise} Resolves: false if there are no events and
+ *    we reached the end of the timeline; else true. Rejects: Error with an error response
+ */
+MatrixClient.prototype.paginateEventTimeline = function(eventTimeline, opts) {
+    // TODO: we should implement a backoff (as per scrollback()) to deal more
+    // nicely with HTTP errors.
+    opts = opts || {};
+    var backwards = opts.backwards || false;
+
+    var token = eventTimeline.getPaginationToken(backwards);
+    if (!token) {
+        // no more results.
+        return q.reject(new Error("No paginate token"));
+    }
+
+    var dir = backwards ? 'b' : 'f';
+    var pendingRequest = eventTimeline._paginationRequests[dir];
+
+    if (pendingRequest) {
+        // already a request in progress - return the existing promise
+        return pendingRequest;
+    }
+
+    var path = utils.encodeUri(
+        "/rooms/$roomId/messages", {$roomId: eventTimeline.getRoomId()}
+    );
+    var params = {
+        from: token,
+        limit: ('limit' in opts) ? opts.limit : 30,
+        dir: dir
+    };
+
+    var self = this;
+
+    var promise =
+        this._http.authedRequest(undefined, "GET", path, params
+    ).then(function(res) {
+        var token = res.end;
+        console.log("client: completed paginate; backwards=" + backwards +
+                    "; token=" + token);
+        var room = self.getRoom(eventTimeline.getRoomId());
```

Comment:
We don't do any safety checks before this to see if we do have a `Room` object for this room ID. This should be null guarded.